eks/ng/autoscaler/cluster-autoscaler.go

```diff
@@ -0,0 +1,412 @@
+        cluster-autoscaler.kubernetes.io/safe-to-evict: 'false'
+    spec:
+      serviceAccountName: cluster-autoscaler
+      containers:
+{{ if ne .ImageURI "" }}{{.ImageURI}}{{ end }}
+          name: cluster-autoscaler
+          resources:
+            limits:
+              cpu: 100m
+              memory: 300Mi
+            requests:
+              cpu: 100m
+              memory: 300Mi
+          command:
+            - ./cluster-autoscaler
+            - --v=4
+            - --stderrthreshold=info
+            - --cloud-provider=aws
+            - --skip-nodes-with-local-storage=false
+            - --expander=least-waste
+{{ if ne .NodeGroupAutoDiscovery "" }}{{.NodeGroupAutoDiscovery}}{{ end }}
+            - --balance-similar-node-groups
+            - --skip-nodes-with-system-pods=false
+          volumeMounts:
+            - name: ssl-certs
+              mountPath: /etc/ssl/certs/ca-certificates.crt
+              readOnly: true
+          imagePullPolicy: "Always"
+      volumes:
+        - name: ssl-certs
+          hostPath:
+            path: "/etc/ssl/certs/ca-bundle.crt"
+`
+
+var images = map[string]string{
+	"1.17": `        - image: us.gcr.io/k8s-artifacts-prod/autoscaling/cluster-autoscaler:v1.17.2`,
+	"1.16": `        - image: us.gcr.io/k8s-artifacts-prod/autoscaling/cluster-autoscaler:v1.16.5`,
+}
+
+const (
+	nodeGroupAuotDiscoveryData      = `            - --node-group-auto-discovery=asg:tag=k8s.io/cluster-autoscaler/enabled,k8s.io/cluster-autoscaler/`
+	clusterAutoScalerDeploymentName = "cluster-autoscaler"
+)
+
+type caSpecData struct {
+	ImageURI               string
+	NodeGroupAutoDiscovery string
+}
+
+// Config defines version upgrade configuration.
+type Config struct {
+	Logger    *zap.Logger
+	Stopc     chan struct{}
+	EKSConfig *eksconfig.Config
+	K8SClient k8s_client.EKS
+	EC2API    ec2iface.EC2API
+	ASGAPI    autoscalingiface.AutoScalingAPI
+	EKSAPI    eksiface.EKSAPI
+}
+
+// ClusterAutoScaler defines cluster autoscaler operation.
+type ClusterAutoscaler interface {
+	Create() error
+}
+
+// New creates a new cluster autoscaler.
+func New(cfg Config) ClusterAutoscaler {
+	cfg.Logger.Info("creating tester", zap.String("tester", reflect.TypeOf(tester{}).PkgPath()))
+	return &tester{cfg: cfg}
+}
+
+type tester struct {
+	cfg Config
+}
+
+func (ts *tester) Create() error {
+	needInstall := false
+	for _, cur := range ts.cfg.EKSConfig.AddOnNodeGroups.ASGs {
+		if cur.ClusterAutoScaler.Enable {
+			needInstall = true
+			break
+		}
+	}
+	if !needInstall {
+		ts.cfg.Logger.Info("no NG enables CA; skipping")
+		return nil
+	}
+	return ts.createCA()
+}
+
+func (ts *tester) createCA() error {
+	ts.cfg.Logger.Info("creating CA using kubectl", zap.String("name", ts.cfg.EKSConfig.Name))
+	for asgName, cur := range ts.cfg.EKSConfig.AddOnNodeGroups.ASGs {
+		createStart := time.Now()
+		ts.cfg.Logger.Info("creating CA", zap.String("name", asgName))
+		if err := ts.installCA(); err != nil {
```

Comment:
`InstallCA` should take `cur.Name` as a parameter, otherwise we install the same plugin x number of times?

Reply:
> `InstallCA` should take `cur.Name` as a parameter, otherwise we install the same plugin x number of times?

I don't think it will install the same plugin. We are only installing once per cluster as long as one NG is enabled.

Reply:
> for asgName, cur := range ts.cfg.EKSConfig.AddOnNodeGroups.ASGs {

still looping through this for-loop?  now that we are moving cluster autoscaler installer out, we can replace `createCA` with `installCA`?

Reply:
Talked offline. We still need per-ASG updates for tagging